### PR TITLE
Add ability to transform input

### PR DESCRIPTION
**Pull request by praateekmahajan**
Right now the library only supports fetching from `tfrecord` files and yielding it to be consumed as a `torch.Tensor`. However there are several use cases when transformation is necessary.

This also helps maintain consistency with the regular `torch.utils.data.Dataset`, where one implements the [transform func inside the `__getitem__` function](https://github.com/pytorch/tutorials/blob/a8035e7b098437a672029499dc4c3e83396fbd36/beginner_source/data_loading_tutorial.py#L152).

One of the other libraries that I have used in the past is petastorm and they have a similar [interface for providing `TransformSpec`.](https://github.com/uber/petastorm#pytorch-api).

Example use cases
1. Normalize features
1. Pad features
1. Save only indexes for categorical values and look them up from a dict to save I/O.

I added code in readme for understanding the use case.

### CODE REVIEW

File: tfrecord/reader.py [hunk 1]
```diff
@@ -90,9 +90,11 @@ def read_records(start_offset=None, end_offset=None):
 
 
 def tfrecord_loader(data_path: str,
-                    index_path: str,
+                    index_path: typing.Union[str, None],
```
Comment (praateekmahajan) [inline tfrecord/reader.py]: Looks like we accept None as a value for the input, so indicating that it's `str` type is deceptive at best.

File: tfrecord/reader.py [hunk 2]
```diff
@@ -121,12 +123,39 @@ def tfrecord_loader(data_path: str,
...
+    transform_func : a callable, default = None
+        A function that takes in the input `features` i.e the dict
+        provided in the description and returns a desirable output.
+        It's useful for having transformation on our input to get
+        desired output.
+
+    removed_fields : list of str, default=None
+        One of the list of keys that we extract from each record
+        but don't want to return it.
+
```
Comment (vahidk) [inline tfrecord/reader.py]: Remove remove_fields. You can simply not specify features that you don't want loaded.

File: tfrecord/reader.py [hunk 3]
```diff
@@ -121,12 +123,39 @@ def tfrecord_loader(data_path: str,
         count. Necessary to evenly split/shard the dataset among many
         workers (i.e. >1).
 
+    transform_func : a callable, default = None
+        A function that takes in the input `features` i.e the dict
+        provided in the description and returns a desirable output.
+        It's useful for having transformation on our input to get
+        desired output.
```
Comment (vahidk) [inline tfrecord/reader.py]: Change transform_func to transform.

File: setup.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 
 setup(
     name="tfrecord",
-    version="1.10",
+    version="1.11",
```
Comment (vahidk) [inline setup.py]: Don't change the version.

File: README.md [hunk 5]
```diff
@@ -74,3 +74,39 @@ loader = tfrecord.tfrecord_loader("/path/to/data.tfrecord", None, {
...
+
+### Transforming input
+
+There are use cases when you might not want to return your input directly after reading from the tfrecord files.
+A trivial example could be to normalize your input, if you haven't saved the normalized feature in the files.
+
+You can easily achieve this by leveraging the `transform_func` argument.
+There might also be use cases to remove a field after performing a transformation, this can be leveraged using the `removed_fields`.
+
+Both the arguments mentioned above are available in all readers including `tfrecord.tfrecord_loader` and `tfrecord.torch.TFRecordDataset`.
```
Comment (vahidk) [inline README.md]: It's too verbose. Change to this: "You can optionally pass a function as `transform` argument to perform post processing of features before returning. This can for example be used to decode images or normalize colors to a certain range.

File: tfrecord/reader.py [hunk 6]
```diff
@@ -90,9 +90,11 @@ def read_records(start_offset=None, end_offset=None):
 
 
 def tfrecord_loader(data_path: str,
-                    index_path: str,
+                    index_path: typing.Union[str, None],
                     description: typing.Union[typing.List[str], typing.Dict[str, str], None] = None,
-                    shard: typing.Optional[typing.Tuple[int, int]] = None
+                    shard: typing.Optional[typing.Tuple[int, int]] = None,
+                    transform_func: typing.Callable[[dict], typing.Any] = None,
```
Comment (vahidk) [inline tfrecord/reader.py]: Don't add this to tfrecord_loader. It should only be available in PyTorch datasets.

File: README.md [hunk 7]
```diff
@@ -74,3 +74,39 @@ loader = tfrecord.tfrecord_loader("/path/to/data.tfrecord", None, {
...
+ 
+```python
+import tfrecord
+
+def transformation(features):
```
Comment (vahidk) [inline README.md]: Change this to an example where you decode an image with opencv. It's more useful.

File: tfrecord/reader.py (hunk 2 above)
Comment (praateekmahajan) [reply tfrecord/reader.py]: We have a use case here that where we just load a feature, perform a transformation but not necessarily want it to be yielded as torch Tensor. 

For simplicity imagine you have a feature called `file_path`.
Inside your `transform` you load the file at that path and return it with the name `image`.
Now the `file_path` is redundant and not required anymore.

Lmk if you still have an objection, I can remove it. 

Another place to look at could be here in [petastorms API](https://github.com/uber/petastorm#pytorch-api), they have something called `removed_fields`.

File: README.md (hunk 5 above)
Comment (praateekmahajan) [reply README.md]: Change it, and let the removed_fields there be for now until we [resolve](https://github.com/vahidk/tfrecord/pull/31#discussion_r396194827).

File: tfrecord/reader.py (hunk 6 above)
Comment (praateekmahajan) [reply tfrecord/reader.py]: I'm not sure if I understand. How we can only add in PyTorch dataset and not to tfrecord_loader?

In PyTorch dataset we yield an iterator which we get from `reader.tfrecord_loader`, so we need to make changes there, right?

File: tfrecord/reader.py (hunk 2 above)
Comment (vahidk) [reply tfrecord/reader.py]: You can simply not return it in transform. Let's remove iit.

File: README.md (hunk 7 above)
Comment (praateekmahajan) [reply README.md]: Can you confirm if this looks good and this what you're suggesting? Haven't worked with images in a while, so don't want to assume. 

```python
import tfrecord
import cv2

def transformation(features):
    # get BGR image from bytes
    features["image"] = cv2.imdecode(features["image"], -1)
    return features


description = {
    "image": "bytes",
}

dataset = tfrecord.torch.TFRecordDataset("/path/to/data.tfrecord",
                                         index_path=None,
                                         transform=transformation)
```

One use case I find extremely beneficial is that padding variable length sequences, so that dataloader works out of the box.

Comment (vahidk) [reply README.md]: Looks good. Just call transformation, decode_image for clarity.

File: tfrecord/reader.py (hunk 6 above)
Comment (vahidk) [reply tfrecord/reader.py]: https://docs.python.org/3/library/functions.html#map

File: tfrecord/reader.py (hunk 2 above)
Comment (praateekmahajan) [reply tfrecord/reader.py]: I'll remove it, just to confirm though, you're saying that the user should be responsible for explicitly removing any keys in `transform` themselves that they don't need anymore?

I believe you're suggesting something like the following

```python
description = {"file_index" = "int"}
def transformation(features):
        features["image"] = img = cv2.imread(f'{features["file_index"]}.png')
        features.pop("file_index", None)   
        return features

dataset = tfrecord.torch.TFRecordDataset("/path/to/data.tfrecord",
                                         index_path=None,
                                         description=description,
                                         transform=transformation)
```

instead of having removed fields?

File: tfrecord/reader.py (hunk 6 above)
Comment (praateekmahajan) [reply tfrecord/reader.py]: Aah, that's a good use of map. Thanks! Updated code.

JW, what's the reason of not having it in `tfrecord_loader` too? I don't use `tfrecord_loader` almost ever, except when I am debugging.

File: tfrecord/torch/dataset.py [hunk 8]
```diff
@@ -109,4 +128,5 @@ def __iter__(self):
             self.data_pattern, self.index_pattern, self.splits, self.description)
         if self.shuffle_queue_size:
             it = iterator_utils.shuffle_iterator(it, self.shuffle_queue_size)
-        return it
+
+        return map(self.transform, it)
```
Comment (vahidk) [inline tfrecord/torch/dataset.py]: I suggest to use if instead of a trivial transform to keep it consistent.

**Comment by praateekmahajan** [comment]
Hey @vahidk  thanks for reviewing. I believe now all changes have been addressed.

It'll be great you have time to address these two comments/questions.
- https://github.com/vahidk/tfrecord/pull/31#discussion_r396200429
- https://github.com/vahidk/tfrecord/pull/31#discussion_r396197917

File: tfrecord/reader.py (hunk 6 above)
Comment (vahidk) [reply tfrecord/reader.py]: Because tfrecord_loader is a low level API.
